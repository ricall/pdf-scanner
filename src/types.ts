export type PdfPage = {
    page: number;
    angle: number;
}

export type PdfMetadata = {
    title?: string,
    author?: string,
    subject?: string,
    creator?: string,
    keywords?: string,
    producer?: string,
    creationDate?: Date,
    modificationDate?: Date,
}

export type PdfDocument = {
    name: string,
    bytes: ArrayBuffer,
    pages: PdfPage[],
    currentPage: number,
    metadata: PdfMetadata,
}

export const initialState: PdfDocument = {
    name: 'untitled',
    bytes: new ArrayBuffer(0),
    pages: [],
    currentPage: 0,
    metadata: {
        title: '',
        author: '',
        subject: '',
        creator: '',
        keywords: '',
        producer: '',
        creationDate: undefined,
        modificationDate: undefined,
    }
}

export type PdfAction =
    | { type: 'LOAD', name: string, document: ArrayBuffer, pages: PdfPage[], currentPage: number, metadata: PdfMetadata }
    | { type: 'NEW' }
    | { type: 'SET_CURRENT_PAGE', page: number }
    | { type: 'ROTATE', page: number }
    | { type: 'MOVE_PAGE_LEFT', page: number }
    | { type: 'MOVE_PAGE_RIGHT', page: number }
    | { type: 'DELETE', page: number }
    | { type: 'UPDATE_METADATA', metadata: PdfMetadata }
