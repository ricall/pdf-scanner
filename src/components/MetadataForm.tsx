import React from "react";
import {
    Button,
    Dialog,
    DialogActions,
    DialogContent,
    DialogContentText,
    DialogTitle,
    Grid,
    TextField
} from "@material-ui/core";
import { makeStyles } from "@material-ui/core/styles";
import { KeyboardDateTimePicker, MuiPickersUtilsProvider } from "@material-ui/pickers";
import { Controller, useForm } from "react-hook-form";
import DateFnsUtils from '@date-io/date-fns';
import 'date-fns';
import { PdfMetadata } from "../types";

const useStyles = makeStyles(theme => ({
    root: {
        padding: theme.spacing(1),
        '& .MuiTextField-root': {
            margin: theme.spacing(0.5),
        },
    },
}));

export interface MetadataFormProps {
    metadata: PdfMetadata,
    onChange: (metadata: PdfMetadata, accept: boolean) => void,
}

export const MetadataForm = ({ metadata, onChange }: MetadataFormProps) => {
    const classes = useStyles();
    const { handleSubmit, control } = useForm({ mode: "onTouched" });

    const onSubmit = (data: any) => onChange(data as PdfMetadata, true);
    const onCancel = () => onChange(metadata, false);

    return (
        <Dialog onClose={onCancel} open={true}>
            <form className={classes.root} onSubmit={handleSubmit(onSubmit)}>
                <DialogTitle id="form-dialog-title">Document Metadata</DialogTitle>
                <DialogContent>
                    <DialogContentText>
                        This meta-data is saved along with the PDF document and is used to provide information about
                        the person and the software that was used to edit the pdf document.
                    </DialogContentText>
                    <Controller name="title"
                                control={control}
                                defaultValue={metadata.title}
                                rules={{ required: 'Title is required' }}
                                render={({ field: { onChange, onBlur, value }, fieldState: { error } }) => (
                                    <TextField
                                        label="Title"
                                        fullWidth
                                        value={value}
                                        onChange={onChange}
                                        onBlur={onBlur}
                                        error={!!error}
                                        helperText={error ? error.message : null}/>
                                )}/>

                    <Controller name="author"
                                control={control}
                                defaultValue={metadata.author}
                                rules={{ required: 'Author is required' }}
                                render={({ field: { onChange, onBlur, value }, fieldState: { error } }) => (
                                    <TextField
                                        label="Author"
                                        fullWidth
                                        value={value}
                                        onChange={onChange}
                                        onBlur={onBlur}
                                        error={!!error}
                                        helperText={error ? error.message : null}/>
                                )}/>

                    <Controller name="subject"
                                control={control}
                                defaultValue={metadata.subject}
                                render={({ field: { onChange, value }, fieldState: { error } }) => (
                                    <TextField
                                        label="Subject"
                                        fullWidth
                                        value={value}
                                        onChange={onChange}
                                        error={!!error}
                                        helperText={error ? error.message : null}/>
                                )}/>

                    <Controller name="creator"
                                control={control}
                                defaultValue={metadata.creator}
                                render={({ field: { onChange, value }, fieldState: { error } }) => (
                                    <TextField
                                        label="Creator"
                                        fullWidth
                                        value={value}
                                        onChange={onChange}
                                        error={!!error}
                                        helperText={error ? error.message : null}/>
                                )}/>

                    <Controller name="keywords"
                                control={control}
                                defaultValue={metadata.keywords}
                                render={({ field: { onChange, value }, fieldState: { error } }) => (
                                    <TextField
                                        label="Keywords"
                                        fullWidth
                                        value={value}
                                        onChange={onChange}
                                        error={!!error}
                                        helperText={error ? error.message : null}/>
                                )}/>

                    <Controller name="producer"
                                control={control}
                                defaultValue={metadata.producer}
                                render={({ field: { onChange, value }, fieldState: { error } }) => (
                                    <TextField
                                        label="Producer"
                                        fullWidth
                                        value={value}
                                        onChange={onChange}
                                        error={!!error}
                                        helperText={error ? error.message : null}/>
                                )}/>
                    <Grid container justifyContent="flex-start">
                        <MuiPickersUtilsProvider utils={DateFnsUtils}>
                            <Controller name="creationDate"
                                        control={control}
                                        defaultValue={metadata.creationDate}
                                        rules={{ required: 'Date is required' }}
                                        render={({ field: { onChange, onBlur, value }, fieldState: { error } }) => (
                                            <KeyboardDateTimePicker
                                                margin="normal"
                                                label="Creation date"
                                                format="dd/MM/yyyy HH:mm:ss"
                                                value={value}
                                                onChange={onChange}
                                                onBlur={onBlur}
                                                error={!!error}
                                                helperText={error ? error.message : null}
                                                KeyboardButtonProps={{
                                                    'aria-label': 'change date',
                                                }}
                                            />
                                        )}/>

                            <Controller name="modificationDate"
                                        control={control}
                                        defaultValue={metadata.modificationDate}
                                        rules={{ required: 'Date is required' }}
                                        render={({ field: { onChange, onBlur, value }, fieldState: { error } }) => (
                                            <KeyboardDateTimePicker

                                                margin="normal"
                                                label="Modification date"
                                                format="dd/MM/yyyy HH:mm:ss"
                                                value={value}
                                                onChange={onChange}
                                                onBlur={onBlur}
                                                error={!!error}
                                                helperText={error ? error.message : null}
                                                KeyboardButtonProps={{
                                                    'aria-label': 'change date',
                                                }}
                                            />
                                        )}/>
                        </MuiPickersUtilsProvider>
                    </Grid>
                </DialogContent>
                <DialogActions>
                    <Button onClick={onCancel} color="primary">
                        Cancel
                    </Button>
                    <Button type="submit" variant="contained" color="primary">
                        Apply Changes
                    </Button>
                </DialogActions>
            </form>
        </Dialog>
    );
}