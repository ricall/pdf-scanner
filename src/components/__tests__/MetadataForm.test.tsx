import React from 'react';
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { MetadataForm } from "../MetadataForm";
import { initialState } from "../../types";

describe('MetadataForm component', () => {

    it('can be rendered', () => {
        const wrapper = shallow(
            <MetadataForm metadata={initialState.metadata} onChange={jest.fn()} />
        );
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();
    });

});
