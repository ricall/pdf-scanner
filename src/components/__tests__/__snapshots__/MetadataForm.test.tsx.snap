// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetadataForm component can be rendered 1`] = `
<WithStyles(ForwardRef(Dialog))
  onClose={[Function]}
  open={true}
>
  <form
    className="makeStyles-root-1"
    onSubmit={[Function]}
  >
    <WithStyles(ForwardRef(DialogTitle))
      id="form-dialog-title"
    >
      Document Metadata
    </WithStyles(ForwardRef(DialogTitle))>
    <WithStyles(ForwardRef(DialogContent))>
      <WithStyles(ForwardRef(DialogContentText))>
        This meta-data is saved along with the PDF document and is used to provide information about the person and the software that was used to edit the pdf document.
      </WithStyles(ForwardRef(DialogContentText))>
      <Component
        control={
          Object {
            "defaultValuesRef": Object {
              "current": Object {},
            },
            "fieldArrayDefaultValuesRef": Object {
              "current": Object {},
            },
            "fieldsRef": Object {
              "current": Object {},
            },
            "formStateRef": Object {
              "current": Object {
                "dirtyFields": Object {},
                "errors": Object {},
                "isDirty": false,
                "isSubmitSuccessful": false,
                "isSubmitted": false,
                "isSubmitting": false,
                "isValid": false,
                "isValidating": false,
                "submitCount": 0,
                "touchedFields": Object {},
              },
            },
            "getIsDirty": [Function],
            "inFieldArrayActionRef": Object {
              "current": false,
            },
            "namesRef": Object {
              "current": Object {
                "array": Set {},
                "mount": Set {},
                "unMount": Set {},
                "watch": Set {},
                "watchAll": false,
              },
            },
            "readFormStateRef": Object {
              "current": Object {
                "dirtyFields": false,
                "errors": false,
                "isDirty": false,
                "isValid": false,
                "isValidating": false,
                "touchedFields": false,
              },
            },
            "register": [Function],
            "shouldUnmount": undefined,
            "subjectsRef": Object {
              "current": Object {
                "array": _e {
                  "observers": Array [],
                },
                "control": _e {
                  "observers": Array [],
                },
                "state": _e {
                  "observers": Array [],
                },
                "watch": _e {
                  "observers": Array [],
                },
              },
            },
            "unregister": [Function],
            "updateIsValid": [Function],
            "watchInternal": [Function],
          }
        }
        defaultValue=""
        name="title"
        render={[Function]}
        rules={
          Object {
            "required": "Title is required",
          }
        }
      />
      <Component
        control={
          Object {
            "defaultValuesRef": Object {
              "current": Object {},
            },
            "fieldArrayDefaultValuesRef": Object {
              "current": Object {},
            },
            "fieldsRef": Object {
              "current": Object {},
            },
            "formStateRef": Object {
              "current": Object {
                "dirtyFields": Object {},
                "errors": Object {},
                "isDirty": false,
                "isSubmitSuccessful": false,
                "isSubmitted": false,
                "isSubmitting": false,
                "isValid": false,
                "isValidating": false,
                "submitCount": 0,
                "touchedFields": Object {},
              },
            },
            "getIsDirty": [Function],
            "inFieldArrayActionRef": Object {
              "current": false,
            },
            "namesRef": Object {
              "current": Object {
                "array": Set {},
                "mount": Set {},
                "unMount": Set {},
                "watch": Set {},
                "watchAll": false,
              },
            },
            "readFormStateRef": Object {
              "current": Object {
                "dirtyFields": false,
                "errors": false,
                "isDirty": false,
                "isValid": false,
                "isValidating": false,
                "touchedFields": false,
              },
            },
            "register": [Function],
            "shouldUnmount": undefined,
            "subjectsRef": Object {
              "current": Object {
                "array": _e {
                  "observers": Array [],
                },
                "control": _e {
                  "observers": Array [],
                },
                "state": _e {
                  "observers": Array [],
                },
                "watch": _e {
                  "observers": Array [],
                },
              },
            },
            "unregister": [Function],
            "updateIsValid": [Function],
            "watchInternal": [Function],
          }
        }
        defaultValue=""
        name="author"
        render={[Function]}
        rules={
          Object {
            "required": "Author is required",
          }
        }
      />
      <Component
        control={
          Object {
            "defaultValuesRef": Object {
              "current": Object {},
            },
            "fieldArrayDefaultValuesRef": Object {
              "current": Object {},
            },
            "fieldsRef": Object {
              "current": Object {},
            },
            "formStateRef": Object {
              "current": Object {
                "dirtyFields": Object {},
                "errors": Object {},
                "isDirty": false,
                "isSubmitSuccessful": false,
                "isSubmitted": false,
                "isSubmitting": false,
                "isValid": false,
                "isValidating": false,
                "submitCount": 0,
                "touchedFields": Object {},
              },
            },
            "getIsDirty": [Function],
            "inFieldArrayActionRef": Object {
              "current": false,
            },
            "namesRef": Object {
              "current": Object {
                "array": Set {},
                "mount": Set {},
                "unMount": Set {},
                "watch": Set {},
                "watchAll": false,
              },
            },
            "readFormStateRef": Object {
              "current": Object {
                "dirtyFields": false,
                "errors": false,
                "isDirty": false,
                "isValid": false,
                "isValidating": false,
                "touchedFields": false,
              },
            },
            "register": [Function],
            "shouldUnmount": undefined,
            "subjectsRef": Object {
              "current": Object {
                "array": _e {
                  "observers": Array [],
                },
                "control": _e {
                  "observers": Array [],
                },
                "state": _e {
                  "observers": Array [],
                },
                "watch": _e {
                  "observers": Array [],
                },
              },
            },
            "unregister": [Function],
            "updateIsValid": [Function],
            "watchInternal": [Function],
          }
        }
        defaultValue=""
        name="subject"
        render={[Function]}
      />
      <Component
        control={
          Object {
            "defaultValuesRef": Object {
              "current": Object {},
            },
            "fieldArrayDefaultValuesRef": Object {
              "current": Object {},
            },
            "fieldsRef": Object {
              "current": Object {},
            },
            "formStateRef": Object {
              "current": Object {
                "dirtyFields": Object {},
                "errors": Object {},
                "isDirty": false,
                "isSubmitSuccessful": false,
                "isSubmitted": false,
                "isSubmitting": false,
                "isValid": false,
                "isValidating": false,
                "submitCount": 0,
                "touchedFields": Object {},
              },
            },
            "getIsDirty": [Function],
            "inFieldArrayActionRef": Object {
              "current": false,
            },
            "namesRef": Object {
              "current": Object {
                "array": Set {},
                "mount": Set {},
                "unMount": Set {},
                "watch": Set {},
                "watchAll": false,
              },
            },
            "readFormStateRef": Object {
              "current": Object {
                "dirtyFields": false,
                "errors": false,
                "isDirty": false,
                "isValid": false,
                "isValidating": false,
                "touchedFields": false,
              },
            },
            "register": [Function],
            "shouldUnmount": undefined,
            "subjectsRef": Object {
              "current": Object {
                "array": _e {
                  "observers": Array [],
                },
                "control": _e {
                  "observers": Array [],
                },
                "state": _e {
                  "observers": Array [],
                },
                "watch": _e {
                  "observers": Array [],
                },
              },
            },
            "unregister": [Function],
            "updateIsValid": [Function],
            "watchInternal": [Function],
          }
        }
        defaultValue=""
        name="creator"
        render={[Function]}
      />
      <Component
        control={
          Object {
            "defaultValuesRef": Object {
              "current": Object {},
            },
            "fieldArrayDefaultValuesRef": Object {
              "current": Object {},
            },
            "fieldsRef": Object {
              "current": Object {},
            },
            "formStateRef": Object {
              "current": Object {
                "dirtyFields": Object {},
                "errors": Object {},
                "isDirty": false,
                "isSubmitSuccessful": false,
                "isSubmitted": false,
                "isSubmitting": false,
                "isValid": false,
                "isValidating": false,
                "submitCount": 0,
                "touchedFields": Object {},
              },
            },
            "getIsDirty": [Function],
            "inFieldArrayActionRef": Object {
              "current": false,
            },
            "namesRef": Object {
              "current": Object {
                "array": Set {},
                "mount": Set {},
                "unMount": Set {},
                "watch": Set {},
                "watchAll": false,
              },
            },
            "readFormStateRef": Object {
              "current": Object {
                "dirtyFields": false,
                "errors": false,
                "isDirty": false,
                "isValid": false,
                "isValidating": false,
                "touchedFields": false,
              },
            },
            "register": [Function],
            "shouldUnmount": undefined,
            "subjectsRef": Object {
              "current": Object {
                "array": _e {
                  "observers": Array [],
                },
                "control": _e {
                  "observers": Array [],
                },
                "state": _e {
                  "observers": Array [],
                },
                "watch": _e {
                  "observers": Array [],
                },
              },
            },
            "unregister": [Function],
            "updateIsValid": [Function],
            "watchInternal": [Function],
          }
        }
        defaultValue=""
        name="keywords"
        render={[Function]}
      />
      <Component
        control={
          Object {
            "defaultValuesRef": Object {
              "current": Object {},
            },
            "fieldArrayDefaultValuesRef": Object {
              "current": Object {},
            },
            "fieldsRef": Object {
              "current": Object {},
            },
            "formStateRef": Object {
              "current": Object {
                "dirtyFields": Object {},
                "errors": Object {},
                "isDirty": false,
                "isSubmitSuccessful": false,
                "isSubmitted": false,
                "isSubmitting": false,
                "isValid": false,
                "isValidating": false,
                "submitCount": 0,
                "touchedFields": Object {},
              },
            },
            "getIsDirty": [Function],
            "inFieldArrayActionRef": Object {
              "current": false,
            },
            "namesRef": Object {
              "current": Object {
                "array": Set {},
                "mount": Set {},
                "unMount": Set {},
                "watch": Set {},
                "watchAll": false,
              },
            },
            "readFormStateRef": Object {
              "current": Object {
                "dirtyFields": false,
                "errors": false,
                "isDirty": false,
                "isValid": false,
                "isValidating": false,
                "touchedFields": false,
              },
            },
            "register": [Function],
            "shouldUnmount": undefined,
            "subjectsRef": Object {
              "current": Object {
                "array": _e {
                  "observers": Array [],
                },
                "control": _e {
                  "observers": Array [],
                },
                "state": _e {
                  "observers": Array [],
                },
                "watch": _e {
                  "observers": Array [],
                },
              },
            },
            "unregister": [Function],
            "updateIsValid": [Function],
            "watchInternal": [Function],
          }
        }
        defaultValue=""
        name="producer"
        render={[Function]}
      />
      <WithStyles(ForwardRef(Grid))
        container={true}
        justifyContent="flex-start"
      >
        <MuiPickersUtilsProvider
          utils={[Function]}
        >
          <Component
            control={
              Object {
                "defaultValuesRef": Object {
                  "current": Object {},
                },
                "fieldArrayDefaultValuesRef": Object {
                  "current": Object {},
                },
                "fieldsRef": Object {
                  "current": Object {},
                },
                "formStateRef": Object {
                  "current": Object {
                    "dirtyFields": Object {},
                    "errors": Object {},
                    "isDirty": false,
                    "isSubmitSuccessful": false,
                    "isSubmitted": false,
                    "isSubmitting": false,
                    "isValid": false,
                    "isValidating": false,
                    "submitCount": 0,
                    "touchedFields": Object {},
                  },
                },
                "getIsDirty": [Function],
                "inFieldArrayActionRef": Object {
                  "current": false,
                },
                "namesRef": Object {
                  "current": Object {
                    "array": Set {},
                    "mount": Set {},
                    "unMount": Set {},
                    "watch": Set {},
                    "watchAll": false,
                  },
                },
                "readFormStateRef": Object {
                  "current": Object {
                    "dirtyFields": false,
                    "errors": false,
                    "isDirty": false,
                    "isValid": false,
                    "isValidating": false,
                    "touchedFields": false,
                  },
                },
                "register": [Function],
                "shouldUnmount": undefined,
                "subjectsRef": Object {
                  "current": Object {
                    "array": _e {
                      "observers": Array [],
                    },
                    "control": _e {
                      "observers": Array [],
                    },
                    "state": _e {
                      "observers": Array [],
                    },
                    "watch": _e {
                      "observers": Array [],
                    },
                  },
                },
                "unregister": [Function],
                "updateIsValid": [Function],
                "watchInternal": [Function],
              }
            }
            name="creationDate"
            render={[Function]}
            rules={
              Object {
                "required": "Date is required",
              }
            }
          />
          <Component
            control={
              Object {
                "defaultValuesRef": Object {
                  "current": Object {},
                },
                "fieldArrayDefaultValuesRef": Object {
                  "current": Object {},
                },
                "fieldsRef": Object {
                  "current": Object {},
                },
                "formStateRef": Object {
                  "current": Object {
                    "dirtyFields": Object {},
                    "errors": Object {},
                    "isDirty": false,
                    "isSubmitSuccessful": false,
                    "isSubmitted": false,
                    "isSubmitting": false,
                    "isValid": false,
                    "isValidating": false,
                    "submitCount": 0,
                    "touchedFields": Object {},
                  },
                },
                "getIsDirty": [Function],
                "inFieldArrayActionRef": Object {
                  "current": false,
                },
                "namesRef": Object {
                  "current": Object {
                    "array": Set {},
                    "mount": Set {},
                    "unMount": Set {},
                    "watch": Set {},
                    "watchAll": false,
                  },
                },
                "readFormStateRef": Object {
                  "current": Object {
                    "dirtyFields": false,
                    "errors": false,
                    "isDirty": false,
                    "isValid": false,
                    "isValidating": false,
                    "touchedFields": false,
                  },
                },
                "register": [Function],
                "shouldUnmount": undefined,
                "subjectsRef": Object {
                  "current": Object {
                    "array": _e {
                      "observers": Array [],
                    },
                    "control": _e {
                      "observers": Array [],
                    },
                    "state": _e {
                      "observers": Array [],
                    },
                    "watch": _e {
                      "observers": Array [],
                    },
                  },
                },
                "unregister": [Function],
                "updateIsValid": [Function],
                "watchInternal": [Function],
              }
            }
            name="modificationDate"
            render={[Function]}
            rules={
              Object {
                "required": "Date is required",
              }
            }
          />
        </MuiPickersUtilsProvider>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogActions))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        onClick={[Function]}
      >
        Cancel
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        type="submit"
        variant="contained"
      >
        Apply Changes
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(DialogActions))>
  </form>
</WithStyles(ForwardRef(Dialog))>
`;
